### PR TITLE
Ruby 23 support

### PR DESCRIPTION
**Pull request by toliman7**
Added new parser for 2_3 changes.
Fixes of single character parsing and regular expression issue in parser 2_0.

### CODE REVIEW

File: bin/generate_parser [hunk 1]
```diff
@@ -3,7 +3,7 @@
 # Run from your JRuby home directory....More smarts needed here.
 
 ###### Change these to tastes ######
-JAY=jay
+JAY=/Users/daniel/Projects/CC/jay/src/jay
```
Comment (enebo) [inline bin/generate_parser]: I would add jay to your path and revert that line

File: nbproject/project.properties [hunk 2]
```diff
@@ -36,10 +36,11 @@ javac.classpath=\
 # Space-separated list of extra javac options
 javac.compilerargs=
 javac.deprecation=false
+javac.external.vm=false
 javac.processorpath=\
     ${javac.classpath}
-javac.source=1.5
-javac.target=1.5
+javac.source=1.7
+javac.target=1.7
```
Comment (enebo) [inline nbproject/project.properties]: Not totally against updating the source level but not sure how important it is if we are not using any newer APIs or language features.

File: src/org/jrubyparser/Parser.java [hunk 3]
```diff
@@ -109,4 +113,16 @@ public void warn(ID id, String fileName, int lineNumber, String message, Object.
...
     }
+    
+    public static void main(String[] argv) throws Exception {
+        ParserConfiguration config = new ParserConfiguration(0, CompatVersion.RUBY2_3);
+        //StringReader reader = new StringReader("@@FabType = $sv.env==\"mtqa4\"? \"f7\":\"f1/8\"");
+        //StringReader reader = new StringReader("h = 343.odd? ? 2: 2");
+        //new Parser().parse("(string)", reader, config);
+        //StringReader reader = new StringReader("?.");
+        FileReader reader = new FileReader("test.rb");
+        new Parser().parse("/Users/daniel/Work/Globalfoundries/qatests_framework/ruby/testcases/siview/Test450_CSScripts.rb", reader, config);
```
Comment (enebo) [inline src/org/jrubyparser/Parser.java]: This should be removed or make more generic

File: src/org/jrubyparser/lexer/Lexer.java [hunk 4]
```diff
@@ -545,7 +545,7 @@ public void reset() {
         resetStacks();
         lex_strterm = null;
         commandStart = true;
-        if (parserSupport != null) isTwoZero = parserSupport.getConfiguration().getVersion().is2_0();
+        if (parserSupport != null) isTwoZero = parserSupport.getConfiguration().getVersion().is2_3();
```
Comment (enebo) [inline src/org/jrubyparser/lexer/Lexer.java]: This seems wrong..I wonder if we need two booleans here is there is something 2.0 or 2.3+?

File: src/org/jrubyparser/lexer/Lexer.java [hunk 5]
```diff
@@ -1606,6 +1610,10 @@ private int ampersand(boolean spaceSeen) throws IOException {
             yaccValue = new Token("&", getPosition());
             setState(LexState.EXPR_BEG);
             return Tokens.tOP_ASGN;
+        case '.':
+            setState(LexState.EXPR_BEG);
+            yaccValue = new Token("&.", getPosition());
+            return Tokens.tLONELY;
```
Comment (enebo) [inline src/org/jrubyparser/lexer/Lexer.java]: This is called Tokens.tANDOT in both MRI and JRuby core.  I suggest renaming it to make comparisons simpler.

File: src/org/jrubyparser/ast/SafeOpAsgnNode.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
+/*
```
Comment (enebo) [inline src/org/jrubyparser/ast/SafeOpAsgnNode.java]: This should add a boolean isLazy like we have in JRuby core.  all call/fcall/vcalls can all be lazy (e.g. lonely).    If we went the route of making safe nodes then we would need to double anything which can be on the right side of the lonely operator.

Comment (toliman7) [inline src/org/jrubyparser/ast/SafeOpAsgnNode.java]: Changed the code to match the JRuby Core

File: src/org/jrubyparser/lexer/Lexer.java (hunk 5 above)
Comment (toliman7) [inline src/org/jrubyparser/lexer/Lexer.java]: Sure, I renamed it to tANDOT

File: src/org/jrubyparser/lexer/Lexer.java (hunk 4 above)
Comment (toliman7) [inline src/org/jrubyparser/lexer/Lexer.java]: corrected the is2_0() method and removed this change.

File: src/org/jrubyparser/Parser.java (hunk 3 above)
Comment (toliman7) [inline src/org/jrubyparser/Parser.java]: Removed

File: nbproject/project.properties (hunk 2 above)
Comment (toliman7) [inline nbproject/project.properties]: 1.5 should be ok

File: bin/generate_parser (hunk 1 above)
Comment (toliman7) [inline bin/generate_parser]: Removed

**Comment by enebo** [comment]
@toliman7 I added VERSIONS = [1.8, 1.9, 2.0, 2.3] to spec helpers and all the defined specs passed...so great job.  If you want to contribute more you can help fill out more positioning specs (in specs).  I might quick try and fix the javadocs so they generate right now.

**Pull request merged**